File: src/db-service/server-messages.ts
// structs and things for messages sent by the server to the client

import { z, literal, object, string, union } from "zod";
import { config_parser as config_struct_parser } from "./config";

export type Ready = z.infer<typeof ready_parser>;
export const ready_parser = object({
	message: literal("ready")
});

export type Ok = z.infer<typeof ok_parser>;
export const ok_parser = object({
	message: literal("ok")
});

export type Error = z.infer<typeof error_parser>;
export const error_parser = object({
	message: literal("error"),
	error: string()
});

export type No = z.infer<typeof no_parser>;
export const no_parser = object({
	message: literal("no")
});

export type Message = z.infer<typeof message_parser>;
export const message_parser = object({
	message: literal("message"),
	id: string(),
	channel_id: string(),
	author_id: string(),
	content: string(),
	attachment_urls: string().array()
});

export type User = z.infer<typeof user_parser>;
export const user_parser = object({
	message: literal("user"),
	id: string(),
	name: string(),
	discriminator: string(),
	avatar_url: string().url()
});

export type Config = z.infer<typeof config_parser>;
export const config_parser = object({
	message: literal("config"),
	config: config_struct_parser
});

export type ServerMessages = z.infer<typeof server_messages>;
export const server_messages = union([
	ready_parser,
	ok_parser,
	error_parser,
	no_parser,
	message_parser,
	user_parser,
	config_parser
]);

File: src/bot/db/server_messages.rs
//! structs and things for messages sent by the server to the client

// https://github.com/serde-rs/serde/issues/760
// workaround for now: https://github.com/serde-rs/serde/issues/760#issuecomment-499570311

use twilight_bot_utils::prelude::*;

use serde::Deserialize;
use serde::de::DeserializeOwned;

pub trait ServerMessage: DeserializeOwned {}

#[derive(Deserialize, Debug)]
pub enum ReadyTag { #[serde(rename = "ready")] Tag }
#[derive(Deserialize, Debug)]
pub struct Ready {
	pub message: ReadyTag
}
impl ServerMessage for Ready {}

#[derive(Deserialize, Debug)]
pub enum OkTag { #[serde(rename = "ok")] Tag }
#[derive(Deserialize, Debug)]
pub struct Ok {
	pub message: OkTag
}
impl ServerMessage for Ok {}

#[derive(Deserialize, Debug)]
pub enum ErrorTag { #[serde(rename = "error")] Tag }
#[derive(Deserialize, Debug)]
pub struct Error {
	pub error: String
}
impl ServerMessage for Error {}

#[derive(Deserialize, Debug)]
pub enum MessageTag { #[serde(rename = "message")] Tag }
#[derive(Deserialize, Debug)]
pub struct Message {
	pub message: MessageTag,
	pub id: Id::<MessageMarker>,
	pub channel_id: Id::<ChannelMarker>,
	pub author_id: Id::<UserMarker>,
	pub content: String,
	pub attachment_urls: Vec<String>
}
impl ServerMessage for Message {}

#[derive(Deserialize, Debug)]
pub enum NoMessageTag { #[serde(rename = "no_message")] Tag }
#[derive(Deserialize, Debug)]
pub struct NoMessage {
	pub message: NoMessageTag
}
impl ServerMessage for NoMessage {}

#[derive(Deserialize, Debug)]
pub enum UserTag { #[serde(rename = "user")] Tag }
#[derive(Deserialize, Debug)]
pub struct User {
	pub message: UserTag,
	pub id: Id::<UserMarker>,
	pub name: String,
	#[serde(with = "super::discriminator")]
	pub discriminator: u16,
	pub avatar_url: String
}

#[derive(Deserialize, Debug)]
pub enum NoUserTag { #[serde(rename = "no_user")] Tag }
#[derive(Deserialize, Debug)]
pub struct NoUser {
	pub message: NoUserTag
}
